milabench/validation/validation.py

```diff
@@ -0,0 +1,158 @@
+import json
+from contextlib import contextmanager
+from dataclasses import dataclass, field
+
+from ..structs import BenchLogEntry
+
+
+class ValidationLayer:
+    """Validation layer interface, captures events, makes report"""
+
+    _return_code = None
+
+    def __init__(self, **kwargs) -> None:
+        # early stop means voir requested milabench to stop the benchmark
+        # this means we can ignore the process return code because it got SIGTERM'ed
+        self.early_stop = False
+
+        # return code of the validation layer
+        # this is used to make milabench fail on critical errors
+        self._return_code = None
+
+    def __call__(self, entry):
+        return self._on_event(entry)
+
+    def _on_event(self, entry: BenchLogEntry):
+        self.early_stop = False
+
+        if entry.event == "stop":
+            self.early_stop = True
+
+        method = getattr(self, f"on_{entry.event}", self.on_event)
+        method(entry)
+
+    def on_stop(self, entry):
+        pass
+
+    def on_config(self, entry):
+        pass
+
+    def on_start(self, entry):
+        pass
+
+    def on_error(self, entry):
+        pass
+
+    def on_line(self, entry):
+        pass
+
+    def on_data(Self, entry):
+        pass
+
+    def on_end(self, entry):
+        pass
+
+    def on_phase(self, entry):
+        pass
+
+    @property
+    def error_code(self):
+        return self._return_code
+
+    def set_error_code(self, code):
+        self._return_code = code
+
+    def end(self):
+        return self._return_code
+
+    def on_event(self, entry: BenchLogEntry):
+        raise NotImplementedError(f"Event on_{entry.event} not implemented")
```

Comment:
Because all known events have default implementations that do nothing, this is only ever going to be called for new events that we add for some reason. I would rename `_on_event` to `on_event` and allow the validator to just override that if they want to do something generic (they can call super() to get the default behavior).

Reply:
done